### PR TITLE
feat(android-cards-view): add how-to-fix CardRow for android 

### PR DESCRIPTION
**Pull request by devaradhanm**
#### Description of changes

Sample screenshot with dummy data
image

#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [ ] Added relevant unit test for your changes. (`yarn test`)
- [ ] Verified code coverage for the changes made. Check coverage report at: `<rootDir>/test-results/unit/coverage`
- [ ] Ran precheckin (`yarn precheckin`)
- [ ] (UI changes only) Added screenshots/GIFs to description above
- [ ] (UI changes only) Verified usability with NVDA/JAWS

### CODE REVIEW

File: src/common/components/cards/how-to-fix-android-card-row.tsx [hunk 1]
```diff
@@ -0,0 +1,82 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+import * as React from 'react';
```
Comment (smoralesd) [inline src/common/components/cards/how-to-fix-android-card-row.tsx]: should this file be under `electron` folder? maybe even `electron\views\automated-checks\components\cards`

File: src/tests/unit/tests/common/components/cards/how-to-fix-android-card-row.test.tsx [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+                howToFixSummary: "This isn't a simple text. It has some words that will be marked as bold.",
+            },
+        };
+
+        const testSubject = shallow(<HowToFixAndroidCardRow {...props} />);
+
+        expect(testSubject.getElement()).toMatchSnapshot();
+    });
+
+    it('renders without matches', () => {
```
Comment (smoralesd) [inline src/tests/unit/tests/common/components/cards/how-to-fix-android-card-row.test.tsx]: should this be `"without formatAsCode property"`??

File: src/common/components/cards/how-to-fix-android-card-row.tsx (hunk 1 above)
Comment (devaradhanm) [reply src/common/components/cards/how-to-fix-android-card-row.tsx]: this will be used by the configuration property, just like how we do for web https://github.com/microsoft/accessibility-insights-web/blob/cda39b427c80b79f26fc4a40c8b165617b3c21fd/src/common/configs/unified-result-property-configurations.tsx#L43

File: src/electron/platform/android/rule-information-provider.ts [hunk 3]
```diff
@@ -92,8 +92,8 @@ export class RuleInformationProvider {
         );
     };
 
-    private buildUnifiedResolution(unformattedText: string, codeStrings: string[] = null): UnifiedResolution {
-        return { howToFixSummary: unformattedText, formatAsCode: codeStrings };
+    private buildUnifiedResolution(unformattedText: string, codeStrings: string[] = null): UnifiedFormatAsCodeResolution {
```
Comment (DaveTryon) [inline src/electron/platform/android/rule-information-provider.ts]: howToFixFormat is not necessarily limited to formatting as code. It could include italics, links, or any other number of things. The type name should not be bound to "FormatAsCode". UnifiedFormattedResolution, perhaps?

Also, the rename should extend to include RuleInformation's GetUnifiedResolutionDelegate type and getUnifiedResolution method

File: src/electron/platform/android/rule-information-provider.ts [hunk 4]
```diff
@@ -92,8 +92,8 @@ export class RuleInformationProvider {
...
+        return { howToFixSummary: unformattedText, howToFixFormat: { howToFix: unformattedText, formatAsCode: codeStrings } };
```
Comment (DaveTryon) [inline src/electron/platform/android/rule-information-provider.ts]: Why are we duplicating the unformattedText in the object? Lisa and I intentionally changed this recently to avoid exactly this behavior, and I'd be very sad to see it return. My guess is that there's downstream code that consumes the howToFixFormat property. I'll argue that the downstream code should consume the UnifiedFormattedResolution (or whatever we end up calling it)--that way it has all of the information that it needs in a single object, with no duplication.

File: src/common/components/cards/how-to-fix-android-card-row.tsx (hunk 1 above)
Comment (smoralesd) [reply src/common/components/cards/how-to-fix-android-card-row.tsx]: I saw the same while working on rendering `text`, `contentDescription` and `className` so I think this is fine until we properly decide what to do with it

File: src/electron/platform/android/rule-information-provider.ts (hunk 3 above)
Comment (devaradhanm) [inline src/electron/platform/android/rule-information-provider.ts]: buildUnifiedFormattedResolution

**Comment by smoralesd** [comment]
2 comments from the screenshot:
- `TouchSizeWcag`'s `how to fix` text seems to be running together (there is not enough space between the words, specially the bolded ones)
- `ColorContrast`'s `how to fix` seems to be missing the colored squares right before the color hex value:
image

I think the first comment should be part of this PR.
I'm not sure about the second one, I guess is a nice to have at this point

File: src/common/components/cards/how-to-fix-android-card-row.tsx [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+    howToFixSplit.forEach((item, index) => {
+        const key = `strong-how-to-fix-${props.index}-${index}`;
+        const content = isNil(item.str) ? item.match : item.str;
+
+        if (content[0] === ' ' && index !== 0) {
+            result.push(<span key={key + '-before-space'}>&nbsp;</span>);
+        }
+
+        result.push(isNil(item.str) ? <Term key={key}>{content}</Term> : <span key={key}>{content}</span>);
```
Comment (smoralesd) [inline src/common/components/cards/how-to-fix-android-card-row.tsx]: 01 - isEmptry vs isNill result on a string
 are you sure you don't want to use `isEmpty` for checking this string values?

File: src/common/types/store-data/unified-data-interface.ts [hunk 6]
```diff
@@ -73,6 +73,15 @@ export type UnifiedResolution = {
     howToFixSummary: string;
 } & InstancePropertyBag;
 
+export type FormattedResolution = {
+    howToFix: string;
+    formatAsCode?: string[];
+};
+
+export type UnifiedFormattedResolution = {
+    howToFixFormat?: FormattedResolution;
```
Comment (smoralesd) [inline src/common/types/store-data/unified-data-interface.ts]: should this be named `howToFixFormatted` or `formattedHowToFix`??? `Format` implies a different meaning I guess

File: src/electron/platform/android/fetch-scan-results.ts [hunk 7]
```diff
@@ -9,7 +10,7 @@ export type HttpGet = typeof axios.get;
 
 export const createFetchScanResults = (httpGet: HttpGet): FetchScanResultsType => {
     return async (port: number) => {
-        const response = await httpGet(`http://localhost:${port}/axe/result`);
-        return new ScanResults(response.data);
+        //  const response = await httpGet(`http://localhost:${port}/axe/result`);
```
Comment (smoralesd) [inline src/electron/platform/android/fetch-scan-results.ts]: comented code here.

File: src/tests/unit/tests/common/components/cards/how-to-fix-android-card-row.test.tsx [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+            howToFix: 'This is a sample string ',
+        },
+        {
+            label: 'when match is on a word',
+            formatAsCode: ['sample'],
+            howToFix: 'This is a sample string ',
+        },
+    ];
+    renderWithSpaceTestCases.forEach(testCase => {
+        it(`renders with whitespace - ${testCase.label}`, () => {
```
Comment (smoralesd) [inline src/tests/unit/tests/common/components/cards/how-to-fix-android-card-row.test.tsx]: is it only because you only want `testCase.label` to show up on the test description console output that you are not using `it.each` here?

File: src/common/components/cards/how-to-fix-android-card-row.tsx [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+import * as React from 'react';
+
+import { Term } from 'assessments/markup';
+import { FormattedResolution } from 'common/types/store-data/unified-data-interface';
+import { isEmpty, isNil } from 'lodash';
+import { CardRowProps } from '../../../common/configs/unified-result-property-configurations';
+import { NamedFC } from '../../../common/react/named-fc';
+import { SimpleCardRow } from './simple-card-row';
+
+export interface HowToFixAndroidCardRowProps extends CardRowProps {
```
Comment (smoralesd) [inline src/common/components/cards/how-to-fix-android-card-row.tsx]: **not for this PR but**: by inheriting from `CardRowProps`, `HowToFixAndroidCardRowProps` contains as a deps a:
```typescript
    fixInstructionProcessor: FixInstructionProcessor;
```
which means, we should be able to add the colored square for color contrast failures.
Again, I don't think we need to do that as part of this PR

File: src/tests/unit/tests/common/components/cards/how-to-fix-android-card-row.test.tsx [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+import { Mock } from 'typemoq';
+
+import { HowToFixAndroidCardRow, HowToFixAndroidCardRowProps } from 'common/components/cards/how-to-fix-android-card-row';
+import { FixInstructionProcessor } from '../../../../../../injected/fix-instruction-processor';
+
+describe(HowToFixAndroidCardRow, () => {
+    it('renders with matches', () => {
+        const fixInstructionProcessorMock = Mock.ofType(FixInstructionProcessor);
+        const props: HowToFixAndroidCardRowProps = {
+            deps: { fixInstructionProcessor: fixInstructionProcessorMock.object },
```
Comment (smoralesd) [inline src/tests/unit/tests/common/components/cards/how-to-fix-android-card-row.test.tsx]: `fixInstructionProcessor` is not really been used on `HowToFixAndroidCardRow`... can't it be null?

File: src/tests/unit/tests/common/components/cards/how-to-fix-android-card-row.test.tsx (hunk 8 above)
Comment (devaradhanm) [reply src/tests/unit/tests/common/components/cards/how-to-fix-android-card-row.test.tsx]: we do have this usage in many places. for eg: content-page.test.tsx

File: src/tests/unit/tests/common/components/cards/how-to-fix-android-card-row.test.tsx (hunk 10 above)
Comment (devaradhanm) [reply src/tests/unit/tests/common/components/cards/how-to-fix-android-card-row.test.tsx]: i will be sending another pr to do the color processing

File: src/common/components/cards/how-to-fix-android-card-row.tsx (hunk 9 above)
Comment (devaradhanm) [reply src/common/components/cards/how-to-fix-android-card-row.tsx]: i will be sending another pr for thi, once this is merged

Comment (devaradhanm) [reply src/common/components/cards/how-to-fix-android-card-row.tsx]: i have added support for color processing now.

File: src/tests/unit/tests/common/components/cards/how-to-fix-android-card-row.test.tsx (hunk 10 above)
Comment (devaradhanm) [reply src/tests/unit/tests/common/components/cards/how-to-fix-android-card-row.test.tsx]: added support for color processing now

File: src/common/components/cards/fix-instruction-color-box.scss [hunk 11]
```diff
@@ -0,0 +1,23 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+@import '../../../common/styles/colors.scss';
```
Comment (devaradhanm) [inline src/common/components/cards/fix-instruction-color-box.scss]: we have duplicated this code everywhere. I have created this new file for styling color box. We need to import this file in all other usages. but, using this new file in other places is not part of this pr.

File: src/electron/platform/android/rule-information-provider.ts [hunk 12]
```diff
@@ -79,21 +79,21 @@ export class RuleInformationProvider {
         return result;
     }
 
-    private getTouchSizeUnifiedResolution = (ruleResultsData: RuleResultsData): UnifiedResolution => {
+    private getTouchSizeUnifiedResolution = (ruleResultsData: RuleResultsData): UnifiedFormattableResolution => {
```
Comment (DaveTryon) [inline src/electron/platform/android/rule-information-provider.ts]: Should this be getTouchSizeUnifiedFormattedResolution instead of getTouchSizeUnifiedResolution?

**Comment by DaveTryon** [comment]
>     private getColorContrastUnifiedResolution = (ruleResultsData: RuleResultsData): UnifiedResolution => {

This should also return a UnfiedFormattableResolution. Rename function to getColorContrasUnifiedFormattableResolution ?

---
Refers to: src/electron/platform/android/rule-information-provider.ts:59 in 27969d4. [](commit_id = 27969d4e541db70ff37746e70df208bff69c0748, deletion_comment = False)

File: src/electron/views/device-connect-view/device-connect-view.scss [hunk 13]
```diff
@@ -2,6 +2,7 @@
 // Licensed under the MIT License.
 @import './components/telemetry-permission-dialog.scss';
 @import '../../../reports/components/outcome.scss';
+@import '../../../common/components/cards/fix-instruction-color-box.scss';
```
Comment (DaveTryon) [inline src/electron/views/device-connect-view/device-connect-view.scss]: Can we remove the `../../..` ?
```suggestion
@import 'common/components/cards/fix-instruction-color-box.scss';
```

File: src/electron/platform/android/rule-information-provider.ts (hunk 4 above)
Comment (DaveTryon) [reply src/electron/platform/android/rule-information-provider.ts]: As we discussed offline, we should have a design conversation to reconcile the differences between issue filing (which operates on the entire `UnifiedResolution` object, and cards, which operates on the `howToFixFormat` object. That's outside the scope of this PR, though, so we'll live with temporary duplication of data

File: src/electron/platform/android/rule-information-provider.ts [hunk 14]
```diff
@@ -1,7 +1,7 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-import { UnifiedResolution } from 'common/types/store-data/unified-data-interface';
+import { UnifiedFormattableResolution, UnifiedResolution } from 'common/types/store-data/unified-data-interface';
```
Comment (DaveTryon) [inline src/electron/platform/android/rule-information-provider.ts]: (Can't add a note on line 59, so adding it here)
All references to `UnifiedResolution` should be replaced with `UnifiedFormattedResolution`. Line 59 is still returning a `UnifiedResolution`. In addition, `getColorContrastUnifiedResolution` might be more appropriate as `getColorContrastUnifiedFormattedResolution`

File: src/common/components/cards/fix-instruction-color-box.scss (hunk 11 above)
Comment (DaveTryon) [inline src/common/components/cards/fix-instruction-color-box.scss]: I think the `../../..` can be omitted:
```suggestion
@import 'common/styles/colors.scss';
```

File: src/electron/platform/android/rule-information.ts [hunk 15]
```diff
@@ -12,12 +12,12 @@ export class RuleInformation {
     constructor(
         readonly ruleId: string,
         readonly ruleDescription: string,
-        readonly getUnifiedResolutionDelegate: GetUnifiedResolutionDelegate,
+        readonly getUnifiedFormattableResolutionDelegate: GetUnifiedResolutionDelegate,
```
Comment (DaveTryon) [inline src/electron/platform/android/rule-information.ts]: Should `GetUnifiedResolutionDelegate` be renamed to `GetUnifiedFormattableResolutionDelegate` ?

File: src/tests/unit/tests/electron/platform/android/rule-information-provider.test.ts [hunk 16]
```diff
@@ -49,7 +49,7 @@ describe('RuleInformationProvider', () => {
 
     function validateUnifiedResolution(ruleId: string, ruleResult: RuleResultsData): UnifiedResolution {
```
Comment (DaveTryon) [inline src/tests/unit/tests/electron/platform/android/rule-information-provider.test.ts]:
```suggestion
    function validateUnifiedResolution(ruleId: string, ruleResult: RuleResultsData): UnifiedFormattableResolution {
```

File: src/tests/unit/tests/electron/platform/android/scan-results-helpers.ts [hunk 17]
```diff
@@ -93,11 +93,11 @@ export function buildRuleInformation(ruleId: string, includeResults: boolean = t
     return {
         ruleId: ruleId,
         ruleDescription: 'This describes ' + ruleId,
-        getUnifiedResolutionDelegate: r => {
+        getUnifiedFormattableResolutionDelegate: r => {
             expect('getUnifiedResolution').toBe('This code should never execute');
```
Comment (DaveTryon) [inline src/tests/unit/tests/electron/platform/android/scan-results-helpers.ts]:
```suggestion
            expect('getUnifiedFormattableResolution').toBe('This code should never execute');
```

**Comment by DaveTryon** [comment]
We should have a chat about how to handle the color sample squares. I'd suggest that we think about adding background and foreground color information to the howToFixFormat object, as opposed to trying to somehow parse it out of the unformatted text. Seems out of scope for this PR either way.

File: src/common/components/cards/fix-instruction-color-box.scss (hunk 11 above)
Comment (smoralesd) [reply src/common/components/cards/fix-instruction-color-box.scss]: that works for typescript file... we haven't done anything like this for scss... don't get me wrong, it may work, but also, it may not

File: src/electron/views/device-connect-view/device-connect-view.scss (hunk 13 above)
Comment (smoralesd) [reply src/electron/views/device-connect-view/device-connect-view.scss]: same as before... works for typescript. not sure about scss
